cmd/keymasterd/awsRole.go

```diff
@@ -0,0 +1,341 @@
+		return nil, fmt.Errorf("verification request failed")
+	}
+	body, err := ioutil.ReadAll(validateResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var callerIdentity getCallerIdentityResponse
+	if err := xml.Unmarshal(body, &callerIdentity); err != nil {
+		return nil, err
+	}
+	parsedArn, err := arn.Parse(callerIdentity.GetCallerIdentityResult.Arn)
+	if err != nil {
+		return nil, err
+	}
+	parsedArn.Region = ""
+	parsedArn.Service = "iam"
+	splitResource := strings.Split(parsedArn.Resource, "/")
+	if len(splitResource) < 2 || splitResource[0] != "assumed-role" {
+		return nil, fmt.Errorf("invalid resource: %s", parsedArn.Resource)
+	}
+	parsedArn.Resource = "role/" + splitResource[1]
+	if parsedArn.String() != claimedArn {
+		return nil, fmt.Errorf("validated ARN: %s != claimed ARN: %s",
+			parsedArn.String(), claimedArn)
+	}
+	return &parsedArnType{
+		parsedArn: parsedArn,
+		role:      splitResource[1],
+	}, nil
+}
+
+func validateStsPresignedUrl(presignedUrl string) error {
+	parsedPresignedUrl, err := url.Parse(presignedUrl)
+	if err != nil {
+		return err
+	}
+	if parsedPresignedUrl.Scheme != "https" {
+		return fmt.Errorf("invalid scheme: %s", parsedPresignedUrl.Scheme)
+	}
+	if parsedPresignedUrl.Path != "/" {
+		return fmt.Errorf("invalid path: %s", parsedPresignedUrl.Path)
+	}
+	if !strings.HasPrefix(parsedPresignedUrl.RawQuery,
+		"Action=GetCallerIdentity&") {
+		return fmt.Errorf("invalid action: %s", parsedPresignedUrl.RawQuery)
+	}
+	splitHost := strings.Split(parsedPresignedUrl.Host, ".")
+	if len(splitHost) != 4 ||
+		splitHost[0] != "sts" ||
+		splitHost[2] != "amazonaws" ||
+		splitHost[3] != "com" {
+		return fmt.Errorf("malformed presigned URL host")
+	}
+	return nil
+}
+
+func (p *assumeRoleCredentialsProvider) Retrieve(ctx context.Context) (
+	aws.Credentials, error) {
+	if time.Until(p.credentials.Expires) > time.Minute {
+		return p.credentials, nil
+	}
+	output, err := p.stsClient.AssumeRole(ctx, &sts.AssumeRoleInput{
+		RoleArn:         p.roleArn,
+		RoleSessionName: aws.String("keymaster"),
+	})
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+	p.credentials = aws.Credentials{
+		AccessKeyID:     *output.Credentials.AccessKeyId,
+		CanExpire:       true,
+		Expires:         *output.Credentials.Expiration,
+		SecretAccessKey: *output.Credentials.SecretAccessKey,
+		SessionToken:    *output.Credentials.SessionToken,
+	}
+	return p.credentials, nil
+}
+
+func (state *RuntimeState) checkAwsRolesEnabled() bool {
+	if len(state.Config.AwsCerts.AllowedAccounts) > 0 {
+		return true
+	}
+	if state.Config.AwsCerts.ListAccountsRole != "" {
+		return true
+	}
+	return false
+}
+
+func (state *RuntimeState) configureAwsRoles() error {
+	if len(state.Config.AwsCerts.AllowedAccounts) > 0 {
+		state.Config.AwsCerts.allowedAccounts =
+			make(map[string]struct{})
+		for _, id := range state.Config.AwsCerts.AllowedAccounts {
+			state.Config.AwsCerts.allowedAccounts[id] = struct{}{}
```

Comment:
please ensure the id is actually a number

Reply:
Done.